src/compute/src/logging/persist.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+use differential_dataflow::Collection;
+use mz_persist_client::write::WriteHandle;
+use mz_repr::GlobalId;
+use mz_storage::controller::CollectionMetadata;
+use mz_storage::types::sources::SourceData;
+use timely::dataflow::channels::pact::Exchange;
+use timely::dataflow::operators::generic::builder_rc::OperatorBuilder;
+use timely::dataflow::Scope;
+use timely::PartialOrder;
+
+use mz_repr::{Diff, Row, Timestamp};
+use mz_timely_util::operators_async_ext::OperatorBuilderExt;
+use timely::progress::Antichain;
+use timely::progress::Timestamp as TimelyTimestamp;
+
+use crate::compute_state::ComputeState;
+
+// TODO(teskje): remove code duplication with `PersistSinkConnection::render_continous_sink`
+pub(crate) fn persist_sink<G>(
+    target_id: GlobalId,
+    target: &CollectionMetadata,
+    compute_state: &mut ComputeState,
+    collection: &Collection<G, Row, Diff>,
+) where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let scope = collection.scope();
+    // TODO(teskje): vary active_worker_index, to distribute work for multiple sinks
+    let active_worker_index = 0;
+
+    let location = target.persist_location.clone();
+
+    let handles: Option<(WriteHandle<_, _, _, _>, _)> = if active_worker_index == scope.index() {
+        let shard_id = target.data_shard;
+        let persist_client = futures_executor::block_on(async {
+            compute_state
+                .persist_clients
+                .lock()
+                .await
+                .open(location)
+                .await
+        });
+
+        let persist_client = persist_client.expect("Successful connection");
+
+        let (write, read) = futures_executor::block_on(
+            persist_client.open::<SourceData, (), Timestamp, Diff>(shard_id),
+        )
+        .expect("could not open persist shard");
```

Comment:
this one is fine. it happens if you reopen the persist shard with different key, val, ts, diff types than when it was created